controllers/machine_controller.go

```diff
@@ -469,22 +469,23 @@ func (r *MachineReconciler) isDeleteNodeAllowed(ctx context.Context, cluster *cl
 		}
 	}
 
-	// Get all of the active machines that belong to this cluster.
-	machines, err := collections.GetFilteredMachinesForCluster(ctx, r.Client, cluster, collections.ActiveMachines)
-	if err != nil {
-		return err
-	}
-
-	// Whether or not it is okay to delete the NodeRef depends on the
-	// number of remaining control plane members and whether or not this
+	// Whether it is okay to delete the NodeRef depends on the
+	// number of remaining control plane members and whether this
 	// machine is one of them.
-	numControlPlaneMachines := len(machines.Filter(collections.ControlPlaneMachines(cluster.Name)))
-	if numControlPlaneMachines == 0 {
-		// Do not delete the NodeRef if there are no remaining members of
-		// the control plane.
-		return errNoControlPlaneNodes
+	if collections.ControlPlaneMachines(cluster.Name)(machine) {
```

Comment:
Please correct me if I'm wrong, but this change greenlights deleting a worker node even if there are no more control plane machines in the Cluster, which is something that will lead to an error preventing a machine to be deleted. 

The current implmementation instead always proceed to delete the machine in the use case above (just skipping node deletion when not possible or not required)

Reply:
+1 the above logic was fixed in a few versions ago. When control plane machines aren't available, there is no way to delete Kubernetes objects in the target cluster, and we should make sure that the Machine controller doesn't get stuck

Reply:
I agree, this should indeed break the `no control plane members` unit test.

Reply:
Oh, sorry. I misunderstood this piece of code logic.

Reply:
So this code comment is outdated, right?

```
// isDeleteNodeAllowed returns nil only if the Machine's NodeRef is not nil
// and if the Machine is not the last control plane node in the cluster.
```